### DISCUSSION TITLE
Looking for an imagemagick pro for a project involving hundreds of scripts

### DISCUSSION THREAD

**Discussion by nicksaroki** [ImageMagick/ImageMagick#3929] [Consulting]
Hey all, I can't give too many details in public, but this is most around automating "renders" of items as though they had been printed on. Please reach out if you have experience!

**Comment by fmw42** [comment]
What platform?  What scripting language?  What version of Imagemagick?

**Reply by nicksaroki** [reply]
Hey Fred! It'd run on Ubuntu. The scripting language doesn't quite matter at this point, it could be shell scripts. Any version of Imagemagick that does the job.

**Comment by snibgo** [comment]
... and what country?

**Reply by nicksaroki** [reply]
US

**Comment by fmw42** [comment]
How does one get more information?

**Comment by nicksaroki** [comment]
I'll start with a generic task... The goal is to scale designs to fit as large as possible within a "mask"

![design](https://user-images.githubusercontent.com/280482/125541378-1d2d9ff4-c703-4830-bdd9-e3001bc2c386.png)
![mask](https://user-images.githubusercontent.com/280482/125541380-e6219a08-00b1-4c52-a87f-c25f437222ee.png)

So the output would be the image from the design scaled to fit inside the circle, centered and as large as possible.

This example is a circle, but the shapes could get more complicated.

**Reply by nicksaroki** [reply]
We tried following this: https://stackoverflow.com/a/49713830 but it was just acting like a normal mask and not scaling.

**Reply by fmw42** [reply]
Did not either my approach or GeeMack's work? Looked like his should work. Working with two aspect ratios can be tricky.

**Reply by fmw42** [reply]
Are both figures above masks or is one the image to go into the mask?

**Reply by nicksaroki** [reply]
The image of the arrow should be resized to fit into the mask.

**Reply by fmw42** [reply]
That does not make much sense since the circle is black on transparency and the arrow is black on transparency. The arrow will not show. Also since the circle has aspect=1, the process is just find the max dimension of the arrow (after a trim) and scale to the dimension of the circle (after trim). Please clarify the issue about the black on transparent for a mask. What do you expect the result to look like?

Are you concerned about clipping of the arrow by the arc of the circle? If so, that become hard for any arbitrary mask shape, even for a circle.

**Reply by nicksaroki** [reply]
Yes, we are concerned about clipping. There can be no clipping at all.

Here is the expected result. It's the design image resized to fit in the space occupied by the "mask":

![result](https://user-images.githubusercontent.com/280482/125544233-1180c92a-c78e-4cce-a570-792d8c714bd6.png)

**Reply by nicksaroki** [reply]
I'm sorry... Github removed the file names. 

design.png = the arrow
mask.png = the circle
result.png = the resized and placed arrow

**Comment by snibgo** [comment]
So we have an abitrary shape (such as an arrow) and a second arbitrary "mask" shape (such as a circle). The task is to find the "-resize" such that the first is as large as possible without exceeding the boundary of the second shape.

Is that correct?

I can see only an iterative method. It is messy.

We can readily find the bounding box of the first image, and the bounding box and an inscribed rectangle for the second image. This gives a maximum and minimum limit for the resize. The best solution is between these two.

If the second shape is geometrically simple (such as a circle or square), the task is much easier.

**Reply by nicksaroki** [reply]
The second shape can get a bit wonky, like an oval or a heart...
